### ISSUE TITLE
Short-circuit evaluation syntax highligting bug

### ISSUE THREAD

**Issue by Johann-Goetzel** [microsoft/pyright#10540]
## Environment data

-   Pylance version: 2025.5.1
-   OS and version: Windows 10 Education 22H2
-   Python version: 3.12.10 (Windows Apps Version)

## Code Snippet

```python
def func():
    print("Function called!")
    return False
print (True and func() and func())
print ("----")
print (True or func())
print ("----")
print (False and func())
```
Output:
```
Function called!
False
----
True
----
False
```

## Repro Steps

Put the code in VS Code and see.

![Image](https://github.com/user-attachments/assets/13d5096c-3fdf-4978-aff8-dab916373f15)

## Expected behavior

The second call of `func()` in `print (True and func() and func())` should not be highlighted as being called but as being not called.

## Actual behavior

The second call is being highlighted as being called.

## Logs

I think thats not necessary.

**Comment by erictraut** [comment]
By "syntax highlighting", are you talking about semantic tokens (the feature that displays identifiers in colors that correspond to their semantic meaning), or are you referring to unused code detection and reporting? Pyright doesn't provide the former; that's a pylance feature. Plus, the semantic meaning of an identifier doesn't change when it is used in a short-circuited expression, so I think pylance is working correctly here.

If you're talking about unreachable code reporting (presented as dimmed gray text), that is provided by pyright. Pyright is working as intended here, so this isn't a bug. Pyright's unreachable code reporting works on the granularity of statements in the Python language. It doesn't report unreachable clauses subexpressions within a statement. Expressions are always analyzed in full — even if portions are short-circuited, and type-related errors are reported for the entire expression. For provably-unreachable statements, pyright does not report any type-related errors; it limits its error reporting to syntax errors only. This approach is consistent with other language servers and type checkers like the one for TypeScript.

**Comment by Johann-Goetzel** [comment]
I don't know how it works, but I noticed that the info when hovering knows that the function returns False, yet the second func() is not grey.

<img width="742" height="214" alt="Image" src="https://github.com/user-attachments/assets/c3548791-c7d5-477a-a6dd-544984fc0c48" />
